### ISSUE TITLE
ol.filter.Crop - can you use a MultiPolygon?

### ISSUE THREAD

**Issue by gjones84** [Viglino/ol-ext#265]
I'm trying to use ol.filter.Crop on a tile layer using about 200 separate polygon shapes - the aim being to have 200 non-contiguous areas on the map in which the tile layer is visible. I tried creating filters (ol.filter.Crop) for each polygon, but I found that while this worked for the first filter, adding a second filter made the image disappear completely - presumably because if there are two filters covering different areas, no part of the TileImage can be within both of those filters at the same time. So I tried just creating one filter using a MultiPolygon instead of a Polygon, but this gives me "jQuery.Deferred exception: myMultiPolygon.getGeometry is not a function".

Is it possible to use a MultiPolygon as the cropping feature for ol.filter.Crop, or is there a way I could "fake it" while not creating 200 separate layers (i.e. a separate cropped layer for each of my polygons)?

**Comment by mike-000** [comment]
It works with the geometry of features, and I can confirm MultiPolygon does work.  If getGeometry isn't a function are you sure you are passing a feature?

**Comment by gjones84** [comment]
Thanks for the reply - I think the initial error was coming from a line from within my own code, but I've removed that and I'm now getting a similar error from inside ol-ext.js. Here is my code:
```
window["mpnCells"] = new ol.geom.MultiPolygon([]);
window["srcCell"].getFeatures().forEach(function(feature) {
    window["mpnCells"].appendPolygon(window["pgn"+feature.get('id')]);
});
window["fltCells"] = new ol.filter.Crop({
    feature: window["mpnCells"],
    inner: false
});
window["lyrRGB_current"].addFilter(window["fltCells"]);
window["fltCells"].set('active', true);
map.addLayer(window["lyrRGB_current"]);
```

The error is as follows:
```
jQuery.Deferred exception: options.feature.getGeometry is not a function TypeError: options.feature.getGeometry is not a function
    at ol.filter.Crop.ol.filter.Mask [as constructor] (https://cdn.rawgit.com/Viglino/ol-ext/master/dist/ol-ext.js:10817:28)
```

The corresponding part of ol-ext.js is as follows - line 10817 is the "switch" statement:
```
ol.filter.Mask = function(options)
{	options = options || {};
	ol.filter.Base.call(this, options);
	if (options.feature)
	{	switch (options.feature.getGeometry().getType())
		{	case "Polygon":
			case "MultiPolygon":
				this.feature_ = options.feature;
				break;
			default: break;
		}
	}
	this.set("inner", options.inner);
	this.fillColor_ = options.fill ? ol.color.asString(options.fill.getColor()) || "rgba(0,0,0,0.2)" : "rgba(0,0,0,0.2)";
}
```

I can confirm that window["mpnCells"] is a MultiPolygon - if I try 'console.log(window["mpnCells"].getType());', I get "MultiPolygon" - but if I try 'console.log(window["mpnCells"].getGeometry().getType());' I get the same error as above. It seems getGeometry() doesn't work on a MultiPolygon. Do I need to do something to my MultiPolygon first before passing it to ol.filter.Crop?

**Comment by ThomasG77** [comment]
Normal, `window["mpnCells"]` does contain a geometry instead of a feature...
So, it make the code from ol-ext failed but it's related to your code, not a bug.
The code below should work better.

```javascript
var multipolygon = new ol.geom.MultiPolygon([]);
window["srcCell"].getFeatures().forEach(function(feature) {
    multiPolygons.appendPolygon(window["pgn"+feature.get('id')]);
});

window["mpnCells"] = new ol.Feature({
  geometry: multipolygon
});

window["fltCells"] = new ol.filter.Crop({
    feature: window["mpnCells"],
    inner: false
});
window["lyrRGB_current"].addFilter(window["fltCells"]);
window["fltCells"].set('active', true);
map.addLayer(window["lyrRGB_current"]);
```

**Comment by gjones84** [comment]
Thanks @ThomasG77 - this definitely seems to be getting me closer. The error has gone away now, but there appears to be something strange happening with rendering - see screen shot below.
![image](https://user-images.githubusercontent.com/12651747/56102554-00bf7000-5f71-11e9-8111-59dccf8895de.png)

I'm expecting my RGB layer (a Tile layer) to be shown within the white boundaries shown on the map - those are my MultiPolygon. There are 200 polygons in that MultiPolygon.

My updated code is as follows - I've added the parts where I'm defining my tile layer just in case that's helpful:
````
		window["srcRGB_current"] = new ol.source.TileImage({
			url: window['urlRGB_current'],
			crossOrigin: 'anonymous',
			tileGrid: L8SourceTileGrid
		});
		window["lyrRGB_current"] = new ol.layer.Tile({
			name: 'lyrRGB_current',
			source: window["srcRGB_current"],
			zIndex: 50,
			visible: true,
			maxZoom: maxZoomLevel,
	        extent: window['farmExtent']
		});
		var mpnCellsMultiPolygon = new ol.geom.MultiPolygon([]);
		window["srcCell"].getFeatures().forEach(function(feature) {
			mpnCellsMultiPolygon.appendPolygon(window["pgn"+feature.get('id')]);
		});
		window["mpnCells"] = new ol.Feature({
		  geometry: mpnCellsMultiPolygon
		});
		window["fltCells"] = new ol.filter.Crop({
			feature: window["mpnCells"],
			inner: false
		});
		window["lyrRGB_current"].addFilter(window["fltCells"]);
		window["fltCells"].set('active', true);
		map.addLayer(window["lyrRGB_current"]);
````

Any ideas as to what I should check next? Thanks again.

**Comment by Viglino** [comment]
I've updated the example with a MultiPolygon to show the result in the examples (http://viglino.github.io/ol-ext/examples/filter/map.filter.crop.html).

**Comment by Viglino** [comment]
Are you sure `mpnCellsMultiPolygon` have good coordinates: make a `console.log(mpnCellsMultiPolygon.getCoordinates())` to see what's inside.
I don't know what is `window["pgn"+feature.get('id')]` but I'd rather get the geom from the features:
````javascript
window["srcCell"].getFeatures().forEach(function(feature) {
  if (feature.getGeometry().getType()==='Polygon') {
    mpnCellsMultiPolygon.appendPolygon(feature.getGeometry());
  }
});
````

**Comment by gjones84** [comment]
I found the problem - my polygons were in XYZ geometry, while the MultiPolygon was only accepting XY geometries through appendPolygon. The result was that my MultiPolygon coordinates ended up looking like this:
```
[x,y]
[z,x]
[y,z] ... and so on
```
I couldn't work out how to make my multiPolygon accept XYZ polygons, so I set my polygons to XY geometry (I don't actually need the Z coordinates) and now all is working perfectly. Thanks for all your help - really appreciate it.
